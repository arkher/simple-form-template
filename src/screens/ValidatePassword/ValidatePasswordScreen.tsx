import { useFormik } from "formik";
import { useCallback, useState } from "react";
import { Button, Container, Header } from "../../components";
import ButtonContainer from "../../components/ButtonContainer/ButtonContainer";
import { sendCredentialsService } from "../../services";
import Email from "./fields/Email";
import Name from "./fields/Name";
import Password from "./fields/Password";
import validationSchema from "./validationSchema";
import PasswordValidationResult from "./fields/PasswordValidationResult";

function BasicData() {
  const formik = useFormik({
    initialValues: {
      name: "",
      email: "",
      password: "",
    },
    onSubmit: (values, actions) => {
      sendCredentials(values, actions.resetForm);
    },
    validationSchema,
  });

  const {
    values,
    touched,
    errors,
    isSubmitting,
    getFieldProps,
    setSubmitting,
    isValid,
  } = formik;

  const buttonDisabled =
    !isValid ||
    !values.name ||
    !values.email ||
    !values.password ||
    isSubmitting;

  const [error, setError] = useState<undefined | boolean>(undefined);

  const sendCredentials = useCallback(
    async (data: IFormData, resetForm: () => void) => {
      try {
        setSubmitting(true);
        const response = await sendCredentialsService.post(data);
        if (response.success) {
          resetForm();
        }
        setError(false);
      } catch (error: unknown) {
        setError(true);
      } finally {
        setSubmitting(false);
      }
    },
    []
  );

  return (
    <Container data-testid="validate-password-screen">
      <Header />

      <form onSubmit={formik.handleSubmit} data-testid="form">
        <div className="flex flex-col justify-center items-center pt-4">
          <Name
            errors={errors}
            disabled={isSubmitting}
            touched={touched}
            {...getFieldProps("name")}
          />
          <Email
            errors={errors}
            disabled={isSubmitting}
            touched={touched}
            {...getFieldProps("email")}
          />
          <Password
            errors={errors}
            disabled={isSubmitting}
            touched={touched}
            {...getFieldProps("password")}
          />
          <PasswordValidationResult password={values.password} />

          <ButtonContainer error={error}>
            <Button
              type="submit"
              name="submit"
              isLoading={isSubmitting}
              disabled={buttonDisabled}
            >
              Enviar
            </Button>
          </ButtonContainer>
        </div>
      </form>
    </Container>
  );
}

export default BasicData;
