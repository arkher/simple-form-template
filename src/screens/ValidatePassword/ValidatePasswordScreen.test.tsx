import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ValidatePasswordScreen from "./ValidatePasswordScreen";
import "@testing-library/jest-dom/extend-expect";

jest.mock("../../constants", () => ({
  API_URL: "www.test.com",
}));

describe("ValidatePasswordScreen", () => {
  it("should render", () => {
    render(<ValidatePasswordScreen />);
    expect(screen.getByTestId("validate-password-screen")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    render(<ValidatePasswordScreen />);
    expect(screen.getByTestId("validate-password-screen")).toMatchSnapshot();
  });

  it("should render the header", () => {
    render(<ValidatePasswordScreen />);
    expect(screen.getByText("Valide sua senha")).toBeTruthy();
  });

  it("should keep button disabled when name input is empty", () => {
    render(<ValidatePasswordScreen />);
    expect(screen.getByTestId("button")).toBeDisabled();
  });

  it("should enable button when all inputs are valid", async () => {
    render(<ValidatePasswordScreen />);
    const user = userEvent.setup();

    await user.type(screen.getByPlaceholderText("Nome"), "Usuario Teste");
    await user.type(screen.getByPlaceholderText("Email"), "test@test.com");
    await user.type(screen.getByPlaceholderText("Senha"), "222222");

    const button = screen.getByRole("button");

    await waitFor(() => expect(button).not.toBeDisabled());
  });
});
