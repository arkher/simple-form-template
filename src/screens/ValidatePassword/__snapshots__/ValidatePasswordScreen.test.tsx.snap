// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatePasswordScreen should match snapshot 1`] = `
<div
  class="h-screen flex flex-col px-4 py-8 md:px-32 md:py-24 bg-white w-screen md:h-auto"
  data-testid="validate-password-screen"
>
  <div
    class="flex flex-col items-start pb-4"
    data-testid="header"
  >
    <h1
      class="text-4xl text-black"
    >
      Valide sua senha
    </h1>
  </div>
  <form>
    <div
      class="flex flex-col justify-center items-center pt-4"
    >
      <input
        class="bg-white mb-8 border-2 border-gray-500 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-4"
        data-testid="input"
        id="name"
        name="name"
        placeholder="Nome"
        type="text"
        value=""
      />
      <input
        class="bg-white mb-8 border-2 border-gray-500 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-4"
        data-testid="input"
        id="email"
        name="email"
        placeholder="Email"
        type="email"
        value=""
      />
      <input
        class="bg-white mb-8 border-2 border-gray-500 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-4"
        data-testid="input"
        id="password"
        name="password"
        placeholder="Senha"
        type="number"
        value=""
      />
      <div
        class="w-full flex flex-1 flex-row items-center justify-end md:mt-16"
        data-testid="button-container"
      >
        <div
          class="flex justify-end"
        >
          <button
            class="absolute bottom-4 left-4 right-4 w-90 h-12 flex border-solid border-1 hover:border-gray-500 focus:border-gray-500 focus:outline-none border-gray-400 justify-center items-center bg-gray-200 text-black rounded-md py-2 px-12 md:w-56 md:relative md:left-0 md:bottom-0 disabled:hover:border-gray-400 disabled:cursor-not-allowed disabled:bg-gray-200  disabled:border-gray-400 disabled:text-gray-500"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Enviar
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
